model/postgres/aws/postgres_resource.rb

```diff
@@ -13,5 +13,20 @@ def aws_upgrade_candidate_server
         .select { |server| PgAwsAmi.where(aws_ami_id: server.vm.boot_image).count > 0 }
         .max_by(&:created_at)
     end
+
+    # The first element is the list of vm_hosts to exclude from the new server,
+    # which is empty for aws servers.
+    # The second element is the list of availability zones to exclude from the new server,
+    # which is the list of all used availability zones.
+    # The third element is the availability zone of the representative server,
+    # which is the availability zone of the new server.
+    def aws_new_server_exclusion_filters
+      exclude_availability_zones, availability_zone = if use_different_az_set?
+        [servers.map { it.vm.nic.nic_aws_resource.subnet_az }.uniq, nil]
```

Comment:
3N+1 query issue. Possible fixes are eager loading:

```suggestion
        [servers(eager: {vm: {nic: :nic_aws_resource}}).map { it.vm.nic.nic_aws_resource.subnet_az }.uniq, nil]
```

or better, getting all information in a single query:

```suggestion
        subnet_azs = NicAwsResource
          .where(id: Nic.where(vm_id: servers_dataset.select(:vm_id).select(:id))
          .select_map(:subnet_az)
        [subnet_azs, nil]
```